File: sine.c
#include <stdio.h>
#include <math.h>
#include <stdlib.h>
#include <string.h>

typedef struct {
    char riff[4];
    int file_size;
    char wave[4];
    char fmt[4];
    int fmt_size;
    short audio_format;
    short num_channels;
    int sample_rate;
    int byte_rate;
    short block_align;
    short bits_per_sample;
    char data[4];
    int data_size;
} WavHeader;

int main() {
    double frequency = 440.0;
    double duration = 1.0;
    int sample_rate = 44100;
    int bits_per_sample = 16;
    int num_channels = 1;
    
    int num_samples = (int)(sample_rate * duration);
    int data_size = num_samples * num_channels * (bits_per_sample / 8);
    int file_size = 36 + data_size;
    
    WavHeader header;
    memcpy(header.riff, "RIFF", 4);
    header.file_size = file_size;
    memcpy(header.wave, "WAVE", 4);
    memcpy(header.fmt, "fmt ", 4);
    header.fmt_size = 16;
    header.audio_format = 1;
    header.num_channels = num_channels;
    header.sample_rate = sample_rate;
    header.byte_rate = sample_rate * num_channels * (bits_per_sample / 8);
    header.block_align = num_channels * (bits_per_sample / 8);
    header.bits_per_sample = bits_per_sample;
    memcpy(header.data, "data", 4);
    header.data_size = data_size;
    
    FILE *fp = fopen("sine.wav", "wb");
    if (!fp) {
        perror("Failed to open file");
        return 1;
    }
    
    fwrite(&header, sizeof(WavHeader), 1, fp);
    
    double amplitude = 16000;
    for (int i = 0; i < num_samples; i++) {
        double sample = amplitude * sin(2.0 * M_PI * frequency * i / sample_rate);
        short s = (short)sample;
        printf("%i\n", s);
        fwrite(&s, sizeof(short), 1, fp);
    }
    
    fclose(fp);
    printf("Created sine.wav: %d Hz, %.1f seconds\n", (int)frequency, duration);
    return 0;
}

File: sine.py
import struct
import math
import sys

def create_sine_wav(filename="sine.wav", frequency=440.0, duration=1.0,
                    sample_rate=44100, bits_per_sample=16, num_channels=1):
    num_samples = int(sample_rate * duration)
    data_size = num_samples * num_channels * (bits_per_sample // 8)
    file_size = 36 + data_size

    with open(filename, "wb") as f:
        f.write(b"RIFF")
        f.write(struct.pack("<I", file_size))
        f.write(b"WAVE")
        f.write(b"fmt ")
        f.write(struct.pack("<I", 16))
        f.write(struct.pack("<H", 1))
        f.write(struct.pack("<H", num_channels))
        f.write(struct.pack("<I", sample_rate))
        f.write(struct.pack("<I", sample_rate * num_channels * (bits_per_sample // 8)))
        f.write(struct.pack("<H", num_channels * (bits_per_sample // 8)))
        f.write(struct.pack("<H", bits_per_sample))
        f.write(b"data")
        f.write(struct.pack("<I", data_size))

        amplitude = 16000
        for i in range(num_samples):
            sample = amplitude * math.sin(2.0 * math.pi * frequency * i / sample_rate)
            s = int(sample)
            f.write(struct.pack("<h", s))

    print(f"Created {filename}: {int(frequency)} Hz, {duration:.1f} seconds")

if __name__ == "__main__":
    create_sine_wav()
